crates/sui-core/src/consensus_validator.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use eyre::WrapErr;
+use fastcrypto::{bls12381::min_sig::BLS12381AggregateSignature, traits::AggregateAuthenticator};
+use std::sync::Arc;
+use tap::TapFallible;
+
+use narwhal_worker::TxValidator;
+use sui_types::{
+    crypto::{
+        AggregateAuthoritySignature, AuthorityPublicKey, AuthoritySignInfoTrait, Signable,
+        VerificationObligation,
+    },
+    message_envelope::Message,
+    messages::{ConsensusTransaction, ConsensusTransactionKind},
+    messages_checkpoint::CheckpointFragmentMessage,
+};
+use tracing::warn;
+
+use crate::authority::AuthorityState;
+
+/// Allows verifying the validity of transactions
+#[derive(Clone)]
+struct ConsensusTxValidator {
+    // a pointer to the Authority state, mostly in order to get access to consensus
+    state: Arc<AuthorityState>,
+}
+
+impl ConsensusTxValidator {
+    #[allow(dead_code)]
+    pub fn new(state: Arc<AuthorityState>) -> Self {
+        Self { state }
+    }
+}
+
+fn tx_from_bytes(tx: &[u8]) -> Result<ConsensusTransaction, eyre::Report> {
+    bincode::deserialize::<ConsensusTransaction>(tx)
+        .wrap_err("Malformed transaction (failed to deserialize)")
+}
+
+impl TxValidator for ConsensusTxValidator {
+    type Error = eyre::Report;
+
+    fn validate(&self, tx: &[u8]) -> Result<(), Self::Error> {
+        let transaction = tx_from_bytes(tx)?;
+        match &transaction.kind {
+            ConsensusTransactionKind::UserTransaction(certificate) => {
+                certificate
+                    .verify_signature(&self.state.committee.load())
+                    .tap_err(|err| {
+                        warn!("Malformed transaction (failed to verify): {err}");
+                    })?;
+            }
+            ConsensusTransactionKind::Checkpoint(fragment) => {
+                fragment.verify().tap_err(|err| {
+                    warn!("Ignoring malformed fragment (failed to verify): {err}");
+                })?;
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_batch(&self, b: &narwhal_types::Batch) -> Result<(), Self::Error> {
+        let txs = b
+            .transactions
+            .iter()
+            .map(|tx| tx_from_bytes(tx))
+            .collect::<Result<Vec<_>, _>>()?;
+        let (fragments, user_transactions): (Vec<_>, Vec<_>) = txs
+            .into_iter()
+            .partition(|tx| matches!(tx.kind, ConsensusTransactionKind::Checkpoint(_)));
+        let committee = self.state.committee.load();
+
+        let mut obligation = VerificationObligation::default();
+        for utx in user_transactions {
+            if let ConsensusTransactionKind::UserTransaction(certificate) = utx.kind {
+                // verify the inner user signature
+                certificate.data().verify()?;
```

Comment:
You could in theory accumulate all user sigs together and batch verify them (whatever can be batched ie only ed25519 sigs)